source/s3_buffer_pool.c

```diff
@@ -0,0 +1,319 @@
+ * Note: current implementation is not optimized and instead tries to be
+ * as straightforward as possible. Given that pool manages a small number
+ * of big allocations, performance impact is not that bad, but something we need
+ * to look into on the next iteration.
+ *
+ * Basic approach is to keep a pool of blocks (block size provided by caller,
+ * ex. 128 MB as a typical size). This is referred to as primary storage.
+ * At high level, each block has offset ptr and allocation basically just means
+ * bumping the offset ptr up and increasing blocks alloc count.
+ * Deallocation decreases the block alloc count and if alloc count ever reaches
+ * 0, the offset ptr is reset back to the start of the block.
+ * For acquires less than a block size, the current approach is to go through
+ * all blocks and see if any of them can fit the size and if not a new block is
+ * created.
+ * For acquires over block size, pool falls back to acquiring/releasing through
+ * base allocator.
+ * If memory limit is set, then acquires will return empty buffers when memory
+ * limit is hit. For primary, that means acquires will fail when new block is
+ * needed, but there is no mem for it. For secondary storage, it tracks mem used
+ * by malloc.
+ */
+
+struct aws_s3_buffer_pool_ticket {
+    size_t size;
+    uint8_t *ptr;
+};
+
+/* Default size for blocks array. Note: this is just for meta info, blocks
+ * themselves are not preallocated s*/
+static size_t s_min_num_blocks = 5;
+
+/* Amount of mem reserved for use outside of buffer pool.
+ * This is an optimistic upper bound on mem used as we dont track it. 
+ * Covers both usage outside of pool, i.e. all allocations done as part of s3
+ * client as well as any allocations overruns due to memory waste in the pool. */
+static const size_t s_buffer_pool_reserved_mem = MB_TO_BYTES(128);
+
+struct aws_s3_buffer_pool {
+    struct aws_allocator *base_allocator;
+    struct aws_mutex mutex;
+
+    size_t block_size;
+    /* size at which allocations should go to secondary */
+    size_t primary_size_cutoff;
+
+    size_t mem_limit;
+
+    size_t primary_allocated;
+    size_t primary_reserved;
+    size_t primary_used;
+
+    size_t secondary_reserved;
+    size_t secondary_used;
+
+    size_t acquire_count;
+    size_t release_count;
+
+    struct aws_array_list blocks;
+};
+
+struct s3_buffer_pool_block {
+    size_t block_size;
+    uint8_t *block_ptr;
+    uint8_t *offset_ptr;
+
+    size_t alloc_count;
+};
+
+struct aws_s3_buffer_pool *aws_s3_buffer_pool_new(
+    struct aws_allocator *allocator,
+    size_t block_size,
+    size_t mem_limit) {
+
+    if (mem_limit < GB_TO_BYTES(1)) {
+        AWS_LOGF_ERROR(AWS_LS_S3_CLIENT, "Failed to initialize buffer pool. Min supported value for Memory Limit is 1GB.");
+        aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+        return NULL;
+    }
+
+    if (block_size != 0 || (block_size < 1024 * 1024 || block_size % 1024 != 0)) {
```

Comment:
why allow block_size to be 0?

Reply:
disables primary indirectly. i mostly used it for testing. but can take it out